src/app/InteractionModelEngine.cpp

```diff
@@ -294,6 +293,15 @@ CHIP_ERROR InteractionModelEngine::OnReadInitialRequest(Messaging::ExchangeConte
     ChipLogDetail(InteractionModel, "Received %s request",
                   aInteractionType == ReadHandler::InteractionType::Subscribe ? "Subscribe" : "Read");
 
+    // Reserve the last ReadHandler for ReadInteraction
+    if (aInteractionType == ReadHandler::InteractionType::Subscribe &&
+        ((CHIP_IM_MAX_NUM_READ_HANDLER - GetNumActiveReadHandlers()) == 1))
+    {
+        ChipLogProgress(InteractionModel, "Reserve the last ReadHandler for IM read Interaction");
+        aStatus = Protocols::InteractionModel::Status::ResourceExhausted;
+        return CHIP_ERROR_NO_MEMORY;
```

Comment:
Shouldn't this return CHIP_NO_ERROR just like the case when we don't find a free handler at all?

Also, shouldn't this block come _after_ we go through and try to shut down any existing subscriptions?  Because that might free up a slot (or even several).

Reply:
yes, should happen after the below for loop

Reply:
((CHIP_IM_MAX_NUM_READ_HANDLER - GetNumActiveReadHandlers()) == 1)), if there are 0 free read handler, the code would continue, and return the CHIP_ERROR_NO_MEMORY. if there are only one free readhandler left, and current request is about subscribe, then we reject immediately, and send ResourceExhausted and CHIP_ERROR_NO_MEMORY error

Reply:
> if there are 0 free read handler, the code would continue, and return the CHIP_ERROR_NO_MEMORY

Where would it do that?   What I see is:
```
    for (auto & readHandler : mReadHandlers)
    {
        if (readHandler.IsFree())
...
        }
    }

    ChipLogProgress(InteractionModel, "no resource for %s interaction",
                    aInteractionType == ReadHandler::InteractionType::Subscribe ? "Subscribe" : "Read");
    aStatus = Protocols::InteractionModel::Status::ResourceExhausted;

    return CHIP_NO_ERROR;
```

Reply:
It might not matter much since the caller actually ignores the return value.... which means maybe we should change this to return void, in a followup?  But for now best to be consistent.

Reply:
i see what you mean....

Reply:
fixed now, and return CHIP_NO_ERROR